### PR TITLE
docs: update documentation for environment storage

### PR DESCRIPTION
**Pull request by Lou1415926**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: site/content/docs/commands/storage-init.en.md [hunk 1]
```diff
@@ -3,17 +3,39 @@
 $ copilot storage init
 ```
 ## What does it do?
-`copilot storage init` creates a new storage resource attached to one of your workloads, accessible from inside your service container via a friendly environment variable. You can specify either *S3*, *DynamoDB* or *Aurora* as the resource type.
+`copilot storage init` creates a new storage resource as addons.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
`copilot storage init` creates new storage resources as addons.
```

File: site/content/docs/commands/storage-init.en.md [hunk 2]
```diff
@@ -3,17 +3,39 @@
...
+
+By default, Copilot follows the "database-per-service" pattern:
+only the service or job that you specify during `storage init` will have access to that storage resource.
+The storage is accessible from inside the service's containers via a friendly environment variable.
+
+!!!note ""
+	However, each product has its own unique situation. If you do need your data storage to be shared by multiple service,
+	you can modify the CloudFormation template that Copilot generates for you to achieve your goal.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
	you can modify the Copilot-created CloudFormation template in order to achieve your goal.
```

File: site/content/docs/commands/storage-init.en.md [hunk 3]
```diff
@@ -3,17 +3,39 @@
...
+
+A storage can be created as a [workload addon](../developing/addons/workload.en.md):
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
A storage resource can be created as a [workload addon](../developing/addons/workload.en.md):
```

File: site/content/docs/commands/storage-init.en.md [hunk 4]
```diff
@@ -3,17 +3,39 @@
...
+it is attached to one of your services or jobs, and is deployed and deleted at the same time as the workload.
+For example, when you run `copilot svc deploy --name api`, the storage resource will be deployed along with "api"
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
For example, when you run `copilot svc deploy --name api`, the resource will be deployed along with "api"
```

File: site/content/docs/commands/storage-init.en.md [hunk 5]
```diff
@@ -3,17 +3,39 @@
...
+to the target environment.
+
+Alternatively, a storage can be created as an [environment addon](../developing/addons/environment.en.md):
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
Alternatively, a storage resource can be created as an [environment addon](../developing/addons/environment.en.md):
```

File: site/content/docs/commands/storage-init.en.md [hunk 6]
```diff
@@ -3,17 +3,39 @@
...
+it is attached to environments, and is deployed and deleted at the same time as an environment.
+For example, when you run `copilot env deploy --name test`, the storage resource will be deployed along with the
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
For example, when you run `copilot env deploy --name test`, the resource will be deployed along with the
```

File: site/content/docs/commands/storage-init.en.md [hunk 7]
```diff
@@ -3,17 +3,39 @@
...
 ```
 Required Flags
+  -l, --lifecycle string      Whether the storage should be created and deleted
+                              at the same time as an workload or an environment.
+                              Must be one of: "workload" or "environment".
   -n, --name string           Name of the storage resource to create.
   -t, --storage-type string   Type of storage to add. Must be one of:
                               "DynamoDB", "S3", "Aurora".
-  -w, --workload string       Name of the service or job to associate with storage.
+  -w, --workload string       Name of the service/job that access the storage.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
  -w, --workload string       Name of the service/job that accesses the storage resource.
```

File: site/content/docs/commands/storage-init.en.md [hunk 8]
```diff
@@ -24,31 +46,51 @@ DynamoDB Flags
...
                                Must be of the format '<keyName>:<dataType>'.
+
 Aurora Serverless Flags
       --engine string               The database engine used in the cluster.
                                     Must be either "MySQL" or "PostgreSQL".
       --initial-db string           The initial database to create in the cluster.
       --parameter-group string      Optional. The name of the parameter group to associate with the cluster.
-      --serverless-version string   Optional. Aurora Serverless version. Must be either "v1" or "v2". (default "v2")
+      --serverless-version string   Optional. Aurora Serverless version.
+                                    Must be either "v1" or "v2". (default "v2")
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
                                    Must be either "v1" or "v2" (default "v2").
```

File: site/content/docs/commands/storage-init.en.md [hunk 9]
```diff
@@ -24,31 +46,51 @@ DynamoDB Flags
...
+
+Others Flags
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
Other Flags
```

File: site/content/docs/commands/storage-init.en.md [hunk 10]
```diff
@@ -24,31 +46,51 @@ DynamoDB Flags
...
+      --add-ingress-from string   The workload that needs access to an
+                                  environment storage. Must be specified with
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
                                  environment storage resource. Must be specified
```

File: site/content/docs/commands/storage-init.en.md [hunk 11]
```diff
@@ -24,31 +46,51 @@ DynamoDB Flags
...
+                                  "--name" and "--storage-type".
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
                                  with "--name" and "--storage-type".
```

File: site/content/docs/commands/storage-init.en.md [hunk 12]
```diff
@@ -24,31 +46,51 @@ DynamoDB Flags
...
 
 ## How can I use it? 
 Create an S3 bucket named "my-bucket" attached to the "frontend" service.
 
 ```console
-$ copilot storage init -n my-bucket -t S3 -w frontend
+$ copilot storage init -n my-bucket -t S3 -w frontend -l workload
 ```
+
+Create an environment S3 bucket named "my-bucket", fronted by the "api" service.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
Create an environment S3 bucket named "my-bucket" fronted by the "api" service.
```

File: site/content/docs/commands/storage-init.en.md [hunk 13]
```diff
@@ -67,13 +109,49 @@ $ copilot storage init \
...
 ```
 
+
 ## What happens under the hood?
-Copilot writes a Cloudformation template specifying the S3 bucket or DDB table to the `addons` dir. When you run `copilot svc deploy`, the CLI merges this template with all the other templates in the addons directory to create a nested stack associated with your service. This nested stack describes all the additional resources you've associated with that service and is deployed wherever your service is deployed. 
+Copilot writes a Cloudformation template specifying the S3 bucket, DDB table, or Aurora Serverless cluster to the `addons` dir. 
+When you run `copilot [svc/job/env] deploy`, the CLI merges this template with all the other templates in the addons 
+directory to create a nested stack associated with your service or environment. 
+This nested stack describes all the [additional resources](../developing/addons/workload.en.md) you've associated with 
+that service or environments, and is deployed wherever your service or environments are deployed.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]: ok??
```suggestion
that service or environment, and is deployed wherever your service or environment is deployed. 
```

File: site/content/docs/developing/storage.en.md [hunk 14]
```diff
@@ -4,28 +4,31 @@ There are two ways to add persistence to Copilot workloads: using [`copilot stor
 
 ## Database and Artifacts
 
-To add a database or S3 bucket to your job or service, simply run [`copilot storage init`](../commands/storage-init.en.md).
+To add a database or S3 bucket to your service, job, or environment, simply run [`copilot storage init`](../commands/storage-init.en.md).
```
Comment (huanjani) [inline site/content/docs/developing/storage.en.md:7]: 🎉

File: site/content/docs/developing/storage.en.md [hunk 15]
```diff
@@ -4,28 +4,31 @@ There are two ways to add persistence to Copilot workloads: using [`copilot stor
...
+The next time you run `copilot deploy -n api`, the bucket will be created, permission to access it will be added to the `api` task role,
+and the name of the bucket will be injected into the `api` container under the environment variable `MY_BUCKET_NAME`.
 
 !!!info
     All names are converted into SCREAMING_SNAKE_CASE based on their use of hyphens or underscores. You can view the environment variables for a given service by running `copilot svc show`.
 
-You can also create a [DynamoDB table](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Introduction.html) using `copilot storage init`. For example, to create the Cloudformation template for a table with a sort key and a local secondary index, you could run the following command.
+You can also create a [DynamoDB table](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Introduction.html) using `copilot storage init`.
+For example, to create the Cloudformation template for a table with a sort key and a local secondary index, you could run the following command.
```
Comment (huanjani) [inline site/content/docs/developing/storage.en.md]:
```suggestion
For example, to create the Cloudformation template for a table with a sort key and a local secondary index, you could run the following command:
```

File: site/content/docs/developing/storage.en.md [hunk 16]
```diff
@@ -45,6 +48,16 @@ To create an [RDS Aurora Serverless v1](https://docs.aws.amazon.com/AmazonRDS/la
 $ copilot storage init -n my-cluster -t Aurora --serverless-version v1
 ```
 
+### Environment storage
+
+The `-l` flag is short for `--lifecycle`. In the examples above, the value to the `-l` flag is `workload`.
+This means that the storages will be created as a service addon or a job addon. The storage will be deployed
```
Comment (huanjani) [inline site/content/docs/developing/storage.en.md]:
```suggestion
This means that the storage resources will be created as a service addon or a job addon. The storage will be deployed
```

File: site/content/docs/developing/storage.en.md [hunk 17]
```diff
@@ -45,6 +48,16 @@ To create an [RDS Aurora Serverless v1](https://docs.aws.amazon.com/AmazonRDS/la
...
+when you run `copilot [svc/job] deploy`, and will be deleted when you run `copilot [svc/job] delete`.
+
+Alternatively, if you want your storage to persist even after you delete the service or the job, you can create
+an environment storage. An environment storage is created as an environment addon: it is deployed when you run
```
Comment (huanjani) [inline site/content/docs/developing/storage.en.md]:
```suggestion
an environment storage resource. An environment storage resource is created as an environment addon: it is deployed when you run
```

File: site/content/docs/commands/storage-init.en.md [hunk 18]
```diff
@@ -3,17 +3,39 @@
 $ copilot storage init
 ```
 ## What does it do?
-`copilot storage init` creates a new storage resource attached to one of your workloads, accessible from inside your service container via a friendly environment variable. You can specify either *S3*, *DynamoDB* or *Aurora* as the resource type.
+`copilot storage init` creates a new storage resource as addons.
+
+By default, Copilot follows the "database-per-service" pattern:
+only the service or job that you specify during `storage init` will have access to that storage resource.
+The storage is accessible from inside the service's containers via a friendly environment variable.
```
Comment (iamhopaul123) [inline site/content/docs/commands/storage-init.en.md]: Maybe elaborate what the friendly env var is?

File: site/content/docs/commands/storage-init.en.md [hunk 19]
```diff
@@ -67,13 +109,49 @@ $ copilot storage init \
...
-there will be two buckets deployed, one in the "test" env and one in the "prod" env, accessible only to the "fe" service in its respective environment. 
+
+This generates a CloudFormation template for an S3 bucket that will be deployed and deleted at the same time as an environment.
+In addition, it will generate a CloudFormation template for the access policy that grants "fe" access
+to the "bucket" resource.
+```console
+$ copilot env deploy --name test
+$ copilot env deploy --name prod
+```
+After running the commands, there will be two buckets deployed, one in the "test" env and one in the "prod" env.
```
Comment (iamhopaul123) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
After running the commands, there will be two buckets deployed in "test" and "prod" env respectively.
```

File: site/content/docs/commands/storage-init.en.md (hunk 18 above)
Comment (Lou1415926) [reply site/content/docs/commands/storage-init.en.md]: changed to 

> only the service or job that you specify during `storage init` will have access to that storage resource.
The storage is accessible from inside the service's containers via a friendly environment variable, that
holds the name of your storage resource or credential information that is required is access the resource.

File: site/content/docs/commands/storage-init.en.md (hunk 8 above)
Comment (Lou1415926) [reply site/content/docs/commands/storage-init.en.md]: This was actually autogenerated by cobra. It appends the `(default "v2")` at the end of the flag description.

I applied this change anyway! However, when people run `copilot storage init --help` they will still see `Must be either "v1" or "v2". (default "v2")` though, and this is sort of not within our control

**Review by iamhopaul123 — APPROVED** [review]

File: site/content/docs/commands/storage-init.en.md [hunk 20]
```diff
@@ -146,7 +147,7 @@ to the "bucket" resource.
 $ copilot env deploy --name test
 $ copilot env deploy --name prod
 ```
-After running the commands, there will be two buckets deployed, one in the "test" env and one in the "prod" env.
+After running the commands, there will be two buckets deployed in "test" and "prod" environment, respectively.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
After running the commands, there will be two buckets deployed, one each in the "test" and "prod" environments.
```

File: site/content/docs/commands/storage-init.en.md [hunk 21]
```diff
@@ -67,13 +110,49 @@ $ copilot storage init \
...
 ```
 
+
 ## What happens under the hood?
-Copilot writes a Cloudformation template specifying the S3 bucket or DDB table to the `addons` dir. When you run `copilot svc deploy`, the CLI merges this template with all the other templates in the addons directory to create a nested stack associated with your service. This nested stack describes all the additional resources you've associated with that service and is deployed wherever your service is deployed. 
+Copilot writes a Cloudformation template specifying the S3 bucket, DDB table, or Aurora Serverless cluster to the `addons` dir. 
+When you run `copilot [svc/job/env] deploy`, the CLI merges this template with all the other templates in the addons 
+directory to create a nested stack associated with your service or environment. 
+This nested stack describes all the [additional resources](../developing/addons/workload.en.md) you've associated with 
+that service or the environment, and is deployed wherever your service or an environment is deployed.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
that service or the environment, and is deployed wherever your service or environment is deployed. 
```

File: site/content/docs/commands/storage-init.en.md [hunk 22]
```diff
@@ -3,17 +3,40 @@
 $ copilot storage init
 ```
 ## What does it do?
-`copilot storage init` creates a new storage resource attached to one of your workloads, accessible from inside your service container via a friendly environment variable. You can specify either *S3*, *DynamoDB* or *Aurora* as the resource type.
+`copilot storage init` creates new storage resources as addons.
+
+By default, Copilot follows the "database-per-service" pattern:
+only the service or job that you specify during `storage init` will have access to that storage resource.
+The storage is accessible from inside the service's containers via a friendly environment variable, that
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
The storage is accessible from inside the service's containers via a friendly environment variable that
```

File: site/content/docs/commands/storage-init.en.md [hunk 23]
```diff
@@ -3,17 +3,40 @@
...
+holds the name of your storage resource or credential information that is required is access the resource.
```
Comment (huanjani) [inline site/content/docs/commands/storage-init.en.md]:
```suggestion
holds the name of your storage resource or credential information for accessing the resource.
```

**Review by huanjani — APPROVED** [review]

**Pull request merged**